### PR TITLE
Fix unintentional input unwrapping for OOP workers

### LINKED ISSUE #2504: [Isolated] Durable activity input binding does not respect JSON converter
### Description
The default mechanism used by durable activities to bind input ignores `System.Text.Json` converters, configured for input type. This results in a mismatch -- when activity is called from an orchestration, its input is first serialized by a custom converter JSON converter. However, when an activity is executed, its input is deserialized according to some other rules, which results in a JSON format mismatch and a lengthy exception.

### Expected behavior
Deserialization of activity input uses the same converters as the serialization process (like it happens in the in-process version with `Newtonsoft.Json`).

### Actual behavior
A wild exception with an unbelievably long stack trace. 

### Relevant source code snippets
<details>
<summary> DTO definition & JSON converter </summary>

```csharp
using System.Text.Json;
using System.Text.Json.Serialization;

namespace AzureOutOfProc;

[JsonConverter(typeof(TestConverter))]
public sealed record MyId(Guid Value);

// Serialize MyId as underlying Guid
public sealed class TestConverter : JsonConverter<MyId>
{
    public override MyId? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return new MyId(JsonSerializer.Deserialize<Guid>(ref reader, options));
    }

    public override void Write(Utf8JsonWriter writer, MyId value, JsonSerializerOptions options)
    {
        JsonSerializer.Serialize(writer, value.Value, options);
    }
}
```

</details>

<details>
<summary> Orchestration & activity </summary>

```csharp
using Microsoft.Azure.Functions.Worker;
using Microsoft.DurableTask;

namespace AzureOutOfProc;

public static class OrchestrationClass
{
    [Function(nameof(MyOrchestration))]
    public static async Task<MyId> MyOrchestration([OrchestrationTrigger]TaskOrchestrationContext context, MyId input)
    {
        var result = await context.CallActivityAsync<MyId>(nameof(MyActivity), input);
        return result;
    }

    
    [Function(nameof(MyActivity))]
    public static async Task<MyId> MyActivity([ActivityTrigger] MyId input)
    {
        await Task.Delay(100);
        return new MyId(Guid.NewGuid());
    }
}
```

</details>

<details>
<summary> Http endpoint for debugging </summary>

```csharp
using System.Net;
using Microsoft.Azure.Functions.Worker;
using Microsoft.Azure.Functions.Worker.Http;
using Microsoft.DurableTask;
using Microsoft.DurableTask.Client;

namespace AzureOutOfProc;

public sealed class HttpEndpoints
{
    [Function(nameof(TestEndpoint))]
    public async Task<HttpResponseData> TestEndpoint(
        [HttpTrigger(AuthorizationLevel.Anonymous, "GET", "/testEndpoint")] HttpRequestData request,
        [DurableClient] DurableTaskClient client,
        CancellationToken cancellationToken
    )
    {
        var id = new MyId(Guid.NewGuid());

        var _ = await client.ScheduleNewOrchestrationInstanceAsync(
            new TaskName(nameof(OrchestrationClass.MyOrchestration)),
            id,
            cancellation: cancellationToken
        );

        return request.CreateResponse(HttpStatusCode.OK);
    }
}
```

</details>

<details>
<summary> Host </summary>

```csharp
using Microsoft.Extensions.Hosting;

await new HostBuilder()
    .ConfigureFunctionsWorkerDefaults()
    .Build()
    .RunAsync();
```

</details>

### Known workarounds
Durable orchestrations seem to handle input correctly (checked by observing how often a custom converter is called when calling orchestration). At this point -- making all activities untyped and passing around serialized JSON? No good solution atm, and this issue is a huge blocker, preventing migration from in-proc to isolated proc azure functions.

### App Details

- **Microsoft.Azure.Functions.Worker.Extensions.DurableTask**: 1.0.2
- **Azure Functions runtime version**: 4
- **Programming language used**: C#
- **dotnet**: 7

### Screenshots
<details>
<summary>Full log with stack trace, quite long! </summary>

```
...\JetBrains.DPA.Runner.exe --handle=4036 --backend-pid=32316 --etw-collect-flags=3 --detach-event-name=dpa.detach.4036 ...\func.exe host start --dotnet-isolated-debug --enable-json-output --json-output-file ...\Rider-AzureFunctions-IsolatedWorker-7452json.pid
Azure Functions Core Tools
Core Tools Version:       4.0.5198 Commit hash: N/A  (64-bit)
Function Runtime Version: 4.21.1.20667
[2023-06-28T15:01:37.046Z] Found ...\AzureOutOfProc.csproj. Using for user secrets file configuration
[2023-06-28T15:01:39.976Z] Azure Functions .NET Worker (PID: 26408) initialized in debug mode. Waiting for debugger to attach...
Functions:
        TestEndpoint: [GET,/TESTENDPOINT] http://localhost:7071/api/TestEndpoint
        MyActivity: activityTrigger
        MyOrchestration: orchestrationTrigger
For detailed output, run func with --verbose flag.
[2023-06-28T15:01:43.398Z] Worker process started and initialized.
[2023-06-28T15:01:44.695Z] Executing 'Functions.TestEndpoint' (Reason='This function was programmatically called via the host APIs.', Id=7fcbdee8-3c1b-4c7f-b541-7a8ff14f9a37)
[2023-06-28T15:01:44.933Z] Host lock lease acquired by instance ID '000000000000000000000000906D69CA'.
[2023-06-28T15:01:51.505Z] Scheduling new MyOrchestration orchestration with instance ID '68c3597301a4403eb822703b7f012286' and 38 bytes of input data.
[2023-06-28T15:01:51.780Z] Executed 'Functions.TestEndpoint' (Succeeded, Id=7fcbdee8-3c1b-4c7f-b541-7a8ff14f9a37, Duration=7104ms)
[2023-06-28T15:01:51.798Z] Executing 'Functions.MyOrchestration' (Reason='(null)', Id=2f2c91dc-5b54-4eeb-9f94-887c71077632)
[2023-06-28T15:01:51.972Z] Executed 'Functions.MyOrchestration' (Succeeded, Id=2f2c91dc-5b54-4eeb-9f94-887c71077632, Duration=198ms)
[2023-06-28T15:01:52.016Z] Executing 'Functions.MyActivity' (Reason='(null)', Id=dd720f9f-4b08-44c1-8e2e-cfdf13ffaed8)
[2023-06-28T15:01:52.084Z] Function 'MyActivity', Invocation id 'dd720f9f-4b08-44c1-8e2e-cfdf13ffaed8': An exception was thrown by the invocation.
[2023-06-28T15:01:52.084Z] Result: Function 'MyActivity', Invocation id 'dd720f9f-4b08-44c1-8e2e-cfdf13ffaed8': An exception was thrown by the invocation.
Exception: Microsoft.Azure.Functions.Worker.FunctionInputConverterException: Error converting 1 input parameters for Function 'MyActivity': Cannot convert input parameter 'input' to type 'AzureOutOfProc.MyId' from type 'System.String'. Error:System.Text.Json.JsonException: 'c' is an invalid start of a value. Path: $ | LineNumber: 0 | BytePositionInLine: 0.
[2023-06-28T15:01:52.085Z]  ---> System.Text.Json.JsonReaderException: 'c' is an invalid start of a value. LineNumber: 0 | BytePositionInLine: 0.
[2023-06-28T15:01:52.085Z]    at System.Text.Json.ThrowHelper.ThrowJsonReaderException(Utf8JsonReader& json, ExceptionResource resource, Byte nextByte, ReadOnlySpan`1 bytes)
[2023-06-28T15:01:52.086Z]    at System.Text.Json.Utf8JsonReader.ConsumeValue(Byte marker)
[2023-06-28T15:01:52.086Z]    at System.Text.Json.Utf8JsonReader.ReadFirstToken(Byte first)
[2023-06-28T15:01:52.087Z]    at System.Text.Json.Utf8JsonReader.ReadSingleSegment()
[2023-06-28T15:01:52.088Z]    at System.Text.Json.Utf8JsonReader.Read()
[2023-06-28T15:01:52.088Z]    at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
[2023-06-28T15:01:52.089Z]    --- End of inner exception stack trace ---
[2023-06-28T15:01:52.090Z]    at System.Text.Json.ThrowHelper.ReThrowWithPath(ReadStack& state, JsonReaderException ex)
[2023-06-28T15:01:52.090Z]    at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
[2023-06-28T15:01:52.091Z]    at System.Text.Json.Serialization.JsonConverter`1.ReadCoreAsObject(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
[2023-06-28T15:01:52.092Z]    at System.Text.Json.JsonSerializer.ReadCore[TValue](Utf8JsonReader& reader, JsonTypeInfo jsonTypeInfo, ReadStack& state)
[2023-06-28T15:01:52.092Z]    at System.Text.Json.JsonSerializer.ContinueDeserialize[TValue](ReadBufferState& bufferState, JsonReaderState& jsonReaderState, ReadStack& readStack, JsonTypeInfo jsonTypeInfo)
[2023-06-28T15:01:52.093Z]    at System.Text.Json.JsonSerializer.ReadFromStreamAsync[TValue](Stream utf8Json, JsonTypeInfo jsonTypeInfo, CancellationToken cancellationToken)
[2023-06-28T15:01:52.093Z]    at Microsoft.Azure.Functions.Worker.Converters.JsonPocoConverter.GetConversionResultFromDeserialization(Byte[] bytes, Type type) in D:\a\_work\1\s\src\DotNetWorker.Core\Converters\JsonPocoConve
rter.cs:line 66
[2023-06-28T15:01:52.094Z]    at Microsoft.Azure.Functions.Worker.Context.Features.DefaultFunctionInputBindingFeature.BindFunctionInputAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Context\Features\
DefaultFunctionInputBindingFeature.cs:line 106
[2023-06-28T15:01:52.095Z]    at Microsoft.Azure.Functions.Worker.Invocation.DefaultFunctionExecutor.ExecuteAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Invocation\DefaultFunctionExecutor.cs:line 4
2
[2023-06-28T15:01:52.097Z]    at Microsoft.Azure.Functions.Worker.OutputBindings.OutputBindingsMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next) in D:\a\_work\1\s\src\DotNetWorker.Core\OutputBinding
s\OutputBindingsMiddleware.cs:line 13
[2023-06-28T15:01:52.097Z]    at Microsoft.Azure.Functions.Worker.Extensions.DurableTask.DurableTaskFunctionsMiddleware.Invoke(FunctionContext functionContext, FunctionExecutionDelegate next) in /_/src/Worker.Extensions.Dur
ableTask/DurableTaskFunctionsMiddleware.cs:line 22
[2023-06-28T15:01:52.098Z]    at Microsoft.Azure.Functions.Worker.FunctionsApplication.InvokeFunctionAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\FunctionsApplication.cs:line 77
Stack:    at Microsoft.Azure.Functions.Worker.Context.Features.DefaultFunctionInputBindingFeature.BindFunctionInputAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Context\Features\DefaultFunctionInput
BindingFeature.cs:line 106
[2023-06-28T15:01:52.099Z]    at Microsoft.Azure.Functions.Worker.Invocation.DefaultFunctionExecutor.ExecuteAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Invocation\DefaultFunctionExecutor.cs:line 4
2
[2023-06-28T15:01:52.099Z]    at Microsoft.Azure.Functions.Worker.OutputBindings.OutputBindingsMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next) in D:\a\_work\1\s\src\DotNetWorker.Core\OutputBinding
s\OutputBindingsMiddleware.cs:line 13
[2023-06-28T15:01:52.100Z]    at Microsoft.Azure.Functions.Worker.Extensions.DurableTask.DurableTaskFunctionsMiddleware.Invoke(FunctionContext functionContext, FunctionExecutionDelegate next) in /_/src/Worker.Extensions.Dur
ableTask/DurableTaskFunctionsMiddleware.cs:line 22
[2023-06-28T15:01:52.101Z]    at Microsoft.Azure.Functions.Worker.FunctionsApplication.InvokeFunctionAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\FunctionsApplication.cs:line 77.
[2023-06-28T15:01:52.104Z] Executed 'Functions.MyActivity' (Failed, Id=dd720f9f-4b08-44c1-8e2e-cfdf13ffaed8, Duration=93ms)
[2023-06-28T15:01:52.105Z] System.Private.CoreLib: Exception while executing function: Functions.MyActivity. System.Private.CoreLib: Result: Failure
Exception: Microsoft.Azure.Functions.Worker.FunctionInputConverterException: Error converting 1 input parameters for Function 'MyActivity': Cannot convert input parameter 'input' to type 'AzureOutOfProc.MyId' from type 'Sys
tem.String'. Error:System.Text.Json.JsonException: 'c' is an invalid start of a value. Path: $ | LineNumber: 0 | BytePositionInLine: 0.
[2023-06-28T15:01:52.106Z]  ---> System.Text.Json.JsonReaderException: 'c' is an invalid start of a value. LineNumber: 0 | BytePositionInLine: 0.
[2023-06-28T15:01:52.106Z]    at System.Text.Json.ThrowHelper.ThrowJsonReaderException(Utf8JsonReader& json, ExceptionResource resource, Byte nextByte, ReadOnlySpan`1 bytes)
[2023-06-28T15:01:52.107Z]    at System.Text.Json.Utf8JsonReader.ConsumeValue(Byte marker)
[2023-06-28T15:01:52.107Z]    at System.Text.Json.Utf8JsonReader.ReadFirstToken(Byte first)
[2023-06-28T15:01:52.108Z]    at System.Text.Json.Utf8JsonReader.ReadSingleSegment()
[2023-06-28T15:01:52.108Z]    at System.Text.Json.Utf8JsonReader.Read()
[2023-06-28T15:01:52.109Z]    at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
[2023-06-28T15:01:52.110Z]    --- End of inner exception stack trace ---
[2023-06-28T15:01:52.110Z]    at System.Text.Json.ThrowHelper.ReThrowWithPath(ReadStack& state, JsonReaderException ex)
[2023-06-28T15:01:52.113Z]    at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
[2023-06-28T15:01:52.113Z]    at System.Text.Json.Serialization.JsonConverter`1.ReadCoreAsObject(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
[2023-06-28T15:01:52.114Z]    at System.Text.Json.JsonSerializer.ReadCore[TValue](Utf8JsonReader& reader, JsonTypeInfo jsonTypeInfo, ReadStack& state)
[2023-06-28T15:01:52.114Z]    at System.Text.Json.JsonSerializer.ContinueDeserialize[TValue](ReadBufferState& bufferState, JsonReaderState& jsonReaderState, ReadStack& readStack, JsonTypeInfo jsonTypeInfo)
[2023-06-28T15:01:52.115Z]    at System.Text.Json.JsonSerializer.ReadFromStreamAsync[TValue](Stream utf8Json, JsonTypeInfo jsonTypeInfo, CancellationToken cancellationToken)
[2023-06-28T15:01:52.115Z]    at Microsoft.Azure.Functions.Worker.Converters.JsonPocoConverter.GetConversionResultFromDeserialization(Byte[] bytes, Type type) in D:\a\_work\1\s\src\DotNetWorker.Core\Converters\JsonPocoConve
rter.cs:line 66
[2023-06-28T15:01:52.116Z]    at Microsoft.Azure.Functions.Worker.Context.Features.DefaultFunctionInputBindingFeature.BindFunctionInputAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Context\Features\
DefaultFunctionInputBindingFeature.cs:line 106
[2023-06-28T15:01:52.117Z]    at Microsoft.Azure.Functions.Worker.Invocation.DefaultFunctionExecutor.ExecuteAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Invocation\DefaultFunctionExecutor.cs:line 4
2
[2023-06-28T15:01:52.117Z]    at Microsoft.Azure.Functions.Worker.OutputBindings.OutputBindingsMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next) in D:\a\_work\1\s\src\DotNetWorker.Core\OutputBinding
s\OutputBindingsMiddleware.cs:line 13
[2023-06-28T15:01:52.118Z]    at Microsoft.Azure.Functions.Worker.Extensions.DurableTask.DurableTaskFunctionsMiddleware.Invoke(FunctionContext functionContext, FunctionExecutionDelegate next) in /_/src/Worker.Extensions.Dur
ableTask/DurableTaskFunctionsMiddleware.cs:line 22
[2023-06-28T15:01:52.119Z]    at Microsoft.Azure.Functions.Worker.FunctionsApplication.InvokeFunctionAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\FunctionsApplication.cs:line 77
[2023-06-28T15:01:52.119Z]    at Microsoft.Azure.Functions.Worker.Handlers.InvocationHandler.InvokeAsync(InvocationRequest request) in D:\a\_work\1\s\src\DotNetWorker.Grpc\Handlers\InvocationHandler.cs:line 88
Stack:    at Microsoft.Azure.Functions.Worker.Context.Features.DefaultFunctionInputBindingFeature.BindFunctionInputAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Context\Features\DefaultFunctionInput
BindingFeature.cs:line 106
[2023-06-28T15:01:52.120Z]    at Microsoft.Azure.Functions.Worker.Invocation.DefaultFunctionExecutor.ExecuteAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Invocation\DefaultFunctionExecutor.cs:line 4
2
[2023-06-28T15:01:52.121Z]    at Microsoft.Azure.Functions.Worker.OutputBindings.OutputBindingsMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next) in D:\a\_work\1\s\src\DotNetWorker.Core\OutputBinding
s\OutputBindingsMiddleware.cs:line 13
[2023-06-28T15:01:52.121Z]    at Microsoft.Azure.Functions.Worker.Extensions.DurableTask.DurableTaskFunctionsMiddleware.Invoke(FunctionContext functionContext, FunctionExecutionDelegate next) in /_/src/Worker.Extensions.Dur
ableTask/DurableTaskFunctionsMiddleware.cs:line 22
[2023-06-28T15:01:52.122Z]    at Microsoft.Azure.Functions.Worker.FunctionsApplication.InvokeFunctionAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\FunctionsApplication.cs:line 77
[2023-06-28T15:01:52.122Z]    at Microsoft.Azure.Functions.Worker.Handlers.InvocationHandler.InvokeAsync(InvocationRequest request) in D:\a\_work\1\s\src\DotNetWorker.Grpc\Handlers\InvocationHandler.cs:line 88.
[2023-06-28T15:01:52.127Z] 68c3597301a4403eb822703b7f012286: Function 'MyActivity (Activity)' failed with an error. Reason: Microsoft.Azure.WebJobs.Host.FunctionInvocationException: Exception while executing function: Funct
ions.MyActivity
[2023-06-28T15:01:52.128Z]  ---> Microsoft.Azure.WebJobs.Script.Workers.Rpc.RpcException: Result: Failure
Exception: Microsoft.Azure.Functions.Worker.FunctionInputConverterException: Error converting 1 input parameters for Function 'MyActivity': Cannot convert input parameter 'input' to type 'AzureOutOfProc.MyId' from type 'Sys
tem.String'. Error:System.Text.Json.JsonException: 'c' is an invalid start of a value. Path: $ | LineNumber: 0 | BytePositionInLine: 0.
[2023-06-28T15:01:52.129Z]  ---> System.Text.Json.JsonReaderException: 'c' is an invalid start of a value. LineNumber: 0 | BytePositionInLine: 0.
[2023-06-28T15:01:52.129Z]    at System.Text.Json.ThrowHelper.ThrowJsonReaderException(Utf8JsonReader& json, ExceptionResource resource, Byte nextByte, ReadOnlySpan`1 bytes)
[2023-06-28T15:01:52.130Z]    at System.Text.Json.Utf8JsonReader.ConsumeValue(Byte marker)
[2023-06-28T15:01:52.130Z]    at System.Text.Json.Utf8JsonReader.ReadFirstToken(Byte first)
[2023-06-28T15:01:52.131Z]    at System.Text.Json.Utf8JsonReader.ReadSingleSegment()
[2023-06-28T15:01:52.131Z]    at System.Text.Json.Utf8JsonReader.Read()
[2023-06-28T15:01:52.132Z]    at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
[2023-06-28T15:01:52.132Z]    --- End of inner exception stack trace ---
[2023-06-28T15:01:52.133Z]    at System.Text.Json.ThrowHelper.ReThrowWithPath(ReadStack& state, JsonReaderException ex)
[2023-06-28T15:01:52.133Z]    at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
[2023-06-28T15:01:52.134Z]    at System.Text.Json.Serialization.JsonConverter`1.ReadCoreAsObject(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
[2023-06-28T15:01:52.134Z]    at System.Text.Json.JsonSerializer.ReadCore[TValue](Utf8JsonReader& reader, JsonTypeInfo jsonTypeInfo, ReadStack& state)
[2023-06-28T15:01:52.135Z]    at System.Text.Json.JsonSerializer.ContinueDeserialize[TValue](ReadBufferState& bufferState, JsonReaderState& jsonReaderState, ReadStack& readStack, JsonTypeInfo jsonTypeInfo)
[2023-06-28T15:01:52.136Z]    at System.Text.Json.JsonSerializer.ReadFromStreamAsync[TValue](Stream utf8Json, JsonTypeInfo jsonTypeInfo, CancellationToken cancellationToken)
[2023-06-28T15:01:52.136Z]    at Microsoft.Azure.Functions.Worker.Converters.JsonPocoConverter.GetConversionResultFromDeserialization(Byte[] bytes, Type type) in D:\a\_work\1\s\src\DotNetWorker.Core\Converters\JsonPocoConve
rter.cs:line 66
[2023-06-28T15:01:52.137Z]    at Microsoft.Azure.Functions.Worker.Context.Features.DefaultFunctionInputBindingFeature.BindFunctionInputAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Context\Features\
DefaultFunctionInputBindingFeature.cs:line 106
[2023-06-28T15:01:52.137Z]    at Microsoft.Azure.Functions.Worker.Invocation.DefaultFunctionExecutor.ExecuteAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Invocation\DefaultFunctionExecutor.cs:line 4
2
[2023-06-28T15:01:52.138Z]    at Microsoft.Azure.Functions.Worker.OutputBindings.OutputBindingsMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next) in D:\a\_work\1\s\src\DotNetWorker.Core\OutputBinding
s\OutputBindingsMiddleware.cs:line 13
[2023-06-28T15:01:52.138Z]    at Microsoft.Azure.Functions.Worker.Extensions.DurableTask.DurableTaskFunctionsMiddleware.Invoke(FunctionContext functionContext, FunctionExecutionDelegate next) in /_/src/Worker.Extensions.Dur
ableTask/DurableTaskFunctionsMiddleware.cs:line 22
[2023-06-28T15:01:52.139Z]    at Microsoft.Azure.Functions.Worker.FunctionsApplication.InvokeFunctionAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\FunctionsApplication.cs:line 77
[2023-06-28T15:01:52.139Z]    at Microsoft.Azure.Functions.Worker.Handlers.InvocationHandler.InvokeAsync(InvocationRequest request) in D:\a\_work\1\s\src\DotNetWorker.Grpc\Handlers\InvocationHandler.cs:line 88
Stack:    at Microsoft.Azure.Functions.Worker.Context.Features.DefaultFunctionInputBindingFeature.BindFunctionInputAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Context\Features\DefaultFunctionInput
BindingFeature.cs:line 106
[2023-06-28T15:01:52.140Z]    at Microsoft.Azure.Functions.Worker.Invocation.DefaultFunctionExecutor.ExecuteAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Invocation\DefaultFunctionExecutor.cs:line 4
2
[2023-06-28T15:01:52.141Z]    at Microsoft.Azure.Functions.Worker.OutputBindings.OutputBindingsMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next) in D:\a\_work\1\s\src\DotNetWorker.Core\OutputBinding
s\OutputBindingsMiddleware.cs:line 13
[2023-06-28T15:01:52.141Z]    at Microsoft.Azure.Functions.Worker.Extensions.DurableTask.DurableTaskFunctionsMiddleware.Invoke(FunctionContext functionContext, FunctionExecutionDelegate next) in /_/src/Worker.Extensions.Dur
ableTask/DurableTaskFunctionsMiddleware.cs:line 22
[2023-06-28T15:01:52.143Z]    at Microsoft.Azure.Functions.Worker.FunctionsApplication.InvokeFunctionAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\FunctionsApplication.cs:line 77
[2023-06-28T15:01:52.145Z]    at Microsoft.Azure.Functions.Worker.Handlers.InvocationHandler.InvokeAsync(InvocationRequest request) in D:\a\_work\1\s\src\DotNetWorker.Grpc\Handlers\InvocationHandler.cs:line 88
[2023-06-28T15:01:52.145Z]    at Microsoft.Azure.WebJobs.Script.Description.WorkerFunctionInvoker.InvokeCore(Object[] parameters, FunctionInvocationContext context) in /_/src/WebJobs.Script/Description/Workers/WorkerFunctio
nInvoker.cs:line 101
[2023-06-28T15:01:52.146Z]    at Microsoft.Azure.WebJobs.Script.Description.FunctionInvokerBase.Invoke(Object[] parameters) in /_/src/WebJobs.Script/Description/FunctionInvokerBase.cs:line 82
[2023-06-28T15:01:52.146Z]    at Microsoft.Azure.WebJobs.Script.Description.FunctionGenerator.Coerce[T](Task`1 src) in /_/src/WebJobs.Script/Description/FunctionGenerator.cs:line 225
[2023-06-28T15:01:52.147Z]    at Microsoft.Azure.WebJobs.Host.Executors.FunctionInvoker`2.InvokeAsync(Object instance, Object[] arguments) in D:\a\_work\1\s\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionInvoker.cs:line
 52
[2023-06-28T15:01:52.147Z]    at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.InvokeWithTimeoutAsync(IFunctionInvoker invoker, ParameterHelper parameterHelper, CancellationTokenSource timeoutTokenSource, Cancella
tionTokenSource functionCancellationTokenSource, Boolean throwOnTimeout, TimeSpan timerInterval, IFunctionInstance instance) in D:\a\_work\1\s\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.cs:line 581
[2023-06-28T15:01:52.148Z]    at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.ExecuteWithWatchersAsync(IFunctionInstanceEx instance, ParameterHelper parameterHelper, ILogger logger, CancellationTokenSource functi
onCancellationTokenSource) in D:\a\_work\1\s\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.cs:line 527
[2023-06-28T15:01:52.149Z]    at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.ExecuteWithLoggingAsync(IFunctionInstanceEx instance, FunctionStartedMessage message, FunctionInstanceLogEntry instanceLogEntry, Param
eterHelper parameterHelper, ILogger logger, CancellationToken cancellationToken) in D:\a\_work\1\s\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.cs:line 306
[2023-06-28T15:01:52.149Z]    --- End of inner exception stack trace ---
[2023-06-28T15:01:52.150Z]    at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.ExecuteWithLoggingAsync(IFunctionInstanceEx instance, FunctionStartedMessage message, FunctionInstanceLogEntry instanceLogEntry, Param
eterHelper parameterHelper, ILogger logger, CancellationToken cancellationToken) in D:\a\_work\1\s\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.cs:line 352
[2023-06-28T15:01:52.150Z]    at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.TryExecuteAsync(IFunctionInstance functionInstance, CancellationToken cancellationToken) in D:\a\_work\1\s\src\Microsoft.Azure.WebJobs
.Host\Executors\FunctionExecutor.cs:line 108. IsReplay: False. State: Failed. HubName: TestHubName. AppName: . SlotName: . ExtensionVersion: 2.9.6. SequenceNumber: 6. TaskEventId: 0
[2023-06-28T15:01:52.182Z] Executing 'Functions.MyOrchestration' (Reason='(null)', Id=5b4eb933-f531-47f0-b0dd-0d881dbb8ad5)
[2023-06-28T15:01:52.300Z] Executed 'Functions.MyOrchestration' (Succeeded, Id=5b4eb933-f531-47f0-b0dd-0d881dbb8ad5, Duration=121ms)
[2023-06-28T15:14:12.293Z] Language Worker Process exited. Pid=26408.
[2023-06-28T15:14:12.294Z] C:\Program Files\dotnet\dotnet.exe exited with code 1 (0x1). .

Process finished with exit code 0.

```
</details>

### If deployed to Azure
Not deployed to Azure yet, tested locally. Debugged by Rider, launched something like
```
func.exe host start --dotnet-isolated-debug --enable-json-output --json-output-file <some_path>
```
Endpoint triggered using browser or curl.

### PR DESCRIPTION
**Pull request by jviau**
This PR fixes an issue where when dispatching JToken values to an out-of-proc worker, we would unwrap the serialized contents, thus causing serialization to fail.

### Issue describing the changes in this PR

resolves #2504, https://github.com/microsoft/durabletask-dotnet/issues/199

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation PR is ready to merge and referenced in `pending_docs.md`
* [ ] My changes **should not** be added to the release notes for the next release
    * [x] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)
* [ ] My changes **do not** require any extra work to be leveraged by OutOfProc SDKs
    * [ ] Otherwise: That work is being tracked here: #issue_or_pr_in_each_sdk
* [x] My changes **do not** change the version of the WebJobs.Extensions.DurableTask package
    * [ ] Otherwise: major or minor version updates are reflected in `/src/Worker.Extensions.DurableTask/AssemblyInfo.cs`
* [x] My changes **do not** add EventIds to our EventSource logs
    * [ ] Otherwise: Ensure the EventIds are within the supported range in our existing Windows infrastructure. You may validate this with a deployed app's telemetry. You may also extend the range by completing a PR such as [this one](https://msazure.visualstudio.com/One/_git/AAPT-Antares-Websites/pullrequest/7463263?_a=files).
* [x] My changes **should** be added to v3.x branch.
    * [ ] Otherwise: This change only applies to Durable Functions v2.x and **will not** be merged to branch v3.x.

### CODE REVIEW

**Comment by jviau** [comment]
TODO: need to see if this is a breaking change for other languages (JS for example)

**Review by davidmrdavid — APPROVED** [review]
LGTM but let's wait for CI

File: src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableActivityContext.cs [hunk 1]
```diff
@@ -129,7 +132,7 @@ internal object GetInput(Type destinationType)
             // MessagePayloadDataConverter to throw an exception. This is a workaround for that case. All other
             // inputs with destination System.String (in-proc: JSON and not JSON; out-of-proc: not-JSON) inputs with
             // destination System.String should cast to JValues and be handled above.)
-            if (destinationType.Equals(typeof(string)))
+            if (this.rawInput)
```
Comment (davidmrdavid) [inline src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableActivityContext.cs:135]: retroactive bug find: this change in the if-condition affects the "legacy" OOProc SDKs - JS, PowerShell, Python. In those cases, the destinationType is `string`, so we should have entered this if-statement, but we no longer do.

Comment (davidmrdavid) [reply src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableActivityContext.cs:135]: Ironically, the comment here describes precisely how this if statement was supposed to be used, but we missed that during review.

**Pull request merged**